### DISCUSSION TITLE
2. Typify parameters in syntax-semantic with colors, units, bounds

### DISCUSSION THREAD

**Discussion by hyunjimoon** [Data4DM/BayesSD#7] [perspectives; sense-making]
I've logged a relevant ideas so far.
## Parameter classification
The following is draft from @Tomfid based on our discussion of data for decision and prior classification (into demand, structural, variation).

### Decision tree
![image](https://user-images.githubusercontent.com/30194633/189944141-13043ae0-6175-4583-a390-356f73f3e288.png)

### Mapping tables
Also, we are merging tables into perhaps two: syntax and information.

Parameter Type | Subtype | Example | Note
-- | -- | -- | --
Structural | Physical | Speed of light, transit time for deliveries, absorption rate for levothyroxine in stomach, e, pi | These have a couple of other dimensions of interest, like, how uncertain is it, and is it a location or scale?
  | Behavioral | Price elasticity, attention to delivery delay, trust in management
Noise | Measurement error | Variance of a blood glucose test, error in counting inventory | Often these are characterized by variance or std dev, but there might be a more complete description, adding distribution type, mean (bias), skewness. So, these might be several parameters each.
  | Driving noise | Disturbance to inventory stock from randomness in sales or unmodeled effects like shrinkage (theft & spoilage)
Prior | Parameter | Mean & variance of belief about price elasticity | Not sure Bayesians are thinking of these as parameters, but they're implemented as parameters (constants) in a Vensim model, so included for completeness.
  | Hyperparameter | Plausible variety in covid transmission rates in different locations, where model has [location] detail

Sdspeak | FrequentistSpeak | BayesSpeak
-- | -- | --
Leverage: narrowly, influence of a parameter on model behavior, or broadly, influence of a policy  on a desired outcome | Influence of a data point on a parameter estimate |  
Behavior: trajectory of a model variable over time, or description of agents' decision making in a model

<img width="1005" alt="image" src="https://user-images.githubusercontent.com/30194633/189944578-f755216b-923a-434e-9886-0a2a64fdc722.png">

### Colored diagrams

- Tom's classification of parameters into disease, community, and policy levers are useful as then we could frame the simulated data experiment as "what data do we need to determine the policy levers or answer binary questions":

> <mark style="background: #ABF7F7A6;">disease character</mark>: ref high risk share, all cause sensitivity, seasonal amplitude, initial infection rate, time to onset, time to infectious, time to onset, time to advance, fraction requiring hospitalization, mild duration, treated fatality rate, untreated fatality multiplier
> <mark style="background: #ADCCFFA6;">community character</mark>: initial population, contact density decline, contact stratification, potential frac population quarantined, available hospital beds per capita
> <mark style="background: #D2B3FFA6;">policy lever</mark>: group isolation, reletive isolation effectiveness presymptoms, state policy switch, behavioral risk reduction, relative behavior change for low risk, fatigue threshold?, log behavior change threshold, behavior change threshold, behavior sensitivity, behavior response time, 

![image](https://user-images.githubusercontent.com/30194633/189945665-fb855fab-8bf1-41bb-a1e6-b54268ec76e6.png)
- [discussion](https://metasd.com/2022/08/climate-catastrophe-loops/#comment-235296) from Tom's blog (Angie's question and Tom's reply) is also relevant:
> 1. using the solid arrow for the same level or integration and dotted arrow for differentiation relation of variables. Examples of dotted arrow (in stock management model) are from supplyline to supply rate and supplyline to supplyline adjustment rate.
1. I think this is an interesting idea. However, for CLDs, I tend to minimize the information encoded, because the point is to communicate with people who don’t know the lingo. If I have to start explaining anything more than basic influence, I feel like there’s a risk of losing the audience.

> 2. coloring variables according to semantic and/or unit:
– semantic: green for estimated parameter (as opposed to assumed parameter)
– unit: red for time (month) , green for level (#), blue for rate (#/month), skyblue for ratio (dmnl).
Complexity of this idea makes it less realistic but I’ve seen at least three scholars using colors (unfortunately without much consistency between scholars and even within one scholar) so the need for color might get some consensus.

2. In CLDs, I generally color by discussion topic or to identify key loops. In models, I often color code for variable types or topics, but I don’t have a consistent style. We’ve been thinking about adding some semantic tags to variables, which could be used by a style sheet to automate this (Vensim & Ventity).

> 3. (most unrealistic) use straight arrow for the same unit and curved for different units. Downside is aesthetics; overlapping arrows and less dynamic-feel as flows are usually associated with curve images (for good reasons). I find 3 useful but perhaps 2 above can be a partial solution.
3. I think the aesthetic downside overrides any benefit here.
As an alternative: what if the diagram coloring could be varied on the fly – a menu command to color by units, tags, variable type, etc.?

### Testing-framed decisions
To alleviate cognitive overload, framing the decision (although continuous in reality) as "tests" i.e. binary yes/no. GAN's 0/1 framework is similar in spirit.
```
Is a 3x spike in cases in a few days a looming problem, or a blip?
Should a sports arena be converted to a field hospital?
Is it worth it to buy enough tests for 10,000/day?
Is contact tracing worth it?
Do masks work?
What happens when you reopen the economy?
```

**Comment by hyunjimoon** [comment]
## 1. P laceholder for latent parameter and unrealized-observable
```
# y: set_data
# theta: set_prior_var, set_prior_struc, set_prior_desire(demand)
# return `stanfile` representing P, p(theta, y) as stanfile (lp to optimize)
```
## 2. D ata block input of hyperparameter and realized observed quantities
```
# D: observable
# PA as data2draws, result as data2draws.sample()
# for `draws2data` P, D is hyperparameter (e.g. dimension (S, M, N, T))
# for `data2draws` P, D is hyperparameter + realized observable (above + \tilde{y_1..N})
```
## 3. A pproximator
```
# PAD as data2draws.sample(data=data_draws2data)
# if P was `draws2data`, `.sample` returns y_1..N
# if P was `data2draws`, `.sample` returns \theta_1..M |\tilde{y_1..N}
```

Q. 
- Largest difference btw PA and PAD? p(theta|y), p(theta|y)
- what would PA be in our context?

**Comment by hyunjimoon** [comment]
Modelers should be able to classify variables as classification is the main model assumption.
Hence diagram tool should support modelers to get the sense of variables' type and relation using colors and curve of arrows. Below is how I developed the original stock management model from Business dynamics textbook (left). As mentioned in section Colored diagrams "(most unrealistic) use straight arrow" of https://github.com/Data4DM/BayesSD/discussions/7#discussion-4388229, one reason I can't give up `ds-str` is due to its perceptive support on linearity and units. Some say for better representation of feedback, curvy loops are needed but there are clear costs and benefits. 

![image](https://user-images.githubusercontent.com/30194633/195329262-cd3a7d95-0627-4f4b-9378-045d6cbaaed7.png)

I wish to ask feedback for the following:
1. Between `ds-curvy` vs `ds-line` which one do you prefer?
2. Does the table which classified the variables according to research type (theoretical (only differential equations and scalar parameter), empirical (real data and random variable parameter), managerial (online decision with policy) makes sense to you? If not, which part? Thanks for @tomfid who discovered every time variable (`wip_adj_time`, `inventory_adj_time`, `backlog_adj_time`, `demand_adj_time`) we estimate in empirical study are in fact what we desire to optimize in managerial study.

All three models are uploaded in [MonthlyModel repo](https://github.com/Data4DM/BayesSD/tree/master/MonthlyModel/10%20mng(chain%20%E2%9B%93)).
@Data4DM/maxdiversity

**Reply by tomfid** [reply]
The WIP Control & Inventory Control loops look better in the Orig and ds-curvy variants. The processes around the backlog look better in ds-str, though I think it's more a function of relocating the block than of straight/curvy lines.

I tend to mix curves and lines - I don't think there's a general dominance of one over the other.

**Reply by hyunjimoon** [reply]
Q3. Regarding the idea of adjusting backlog (with the added option of nonzero desired backlog) which has a better naming for current `desired backlog adjust rate` between `desired shipment adjust rate` vs `desired backlog adjust rate`?
Q4. For each stock (WIP vs Inventory vs Backlog), how much would you attribute your attention? Or should we care about their flows? e.g. matching `expected customer order rate` with `production rate` or `production start rate`? As traced in #54 the following is `production_start_rate` initialization:

```
    = max(0, 
        (+ manufacturing_cycle_time * max(0, customer_order_rate + (desired_inventory - desired_inventory)/inventory_adjustment_time) 
         - manufacturing_cycle_time * max(0, customer_order_rate + (desired_inventory - desired_inventory)/inventory_adjustment_time)
        )/wip_adjustment_time
        + max(0, 
              customer_order_rate + 
               (max(0,(+ customer_order_rate * (minimum_order_processing_time + safety_stock_coverage) 
                   -  max(0, customer_order_rate * (minimum_order_processing_time + safety_stock_coverage)
                )/inventory_adjustment_time
            )
      )
```

**Reply by tomfid** [reply]
I think the "Desired Shipment Rate" in the original is the preferred name.

**Reply by tomfid** [reply]
When you say "attribute attention" is that attention in calibration or attention in the stock management policy?

**Reply by tomfid** [reply]
Doesn't the entire term,
(+ manufacturing_cycle_time * max(0, customer_order_rate + (desired_inventory - desired_inventory)/inventory_adjustment_time) 
         - manufacturing_cycle_time * max(0, customer_order_rate + (desired_inventory - desired_inventory)/inventory_adjustment_time)
        )/wip_adjustment_time
cancel out?

I'm not a fan of long equations, especially nested logicals. I sometimes write them anyway, so I guess that means I'm really just not a fan of other people's long equations. Either way, better to break things up into named components if possible.

**Reply by tomfid** [reply]
Production start rate is a flow, so why does it have an initialization? Is it subsequently used to initialize the stocks downstream in equilibrium?

**Reply by hyunjimoon** [reply]
> Production start rate is a flow, so why does it have an initialization?

Perhaps "stocked flow" structure. May I ask whether [this](https://github.com/Data4DM/stanify/blob/main/vignette/stan_files/inven_allpn_S1N3Q2R2_M1000_ps10/functions.stan#L143) final version translated from left below, match your mental model? I am using this for tons of experiments..some verification would be very helpful.

I abstracted different supply chain versions above into right below, but not sure about what and how many to have as target simulated variable.
![image](https://user-images.githubusercontent.com/30194633/207318582-e056404e-2197-448a-b120-e1e09d2b0de1.png)

**Comment by tomfid** [comment]
We just discussed creating a table of parameter types commonly found in SD models, with a recipe for prior selection for each type.

I'm thinking it might be easiest to just co-edit as a .md file. Any opinion about where it should go in the folder structure here? Perhaps we should start a "cookbook" directory containing such things? (Another good candidate for such a page would be the Rosetta Stone table.)

**Reply by hyunjimoon** [reply]
During the exploration period, how about using google sheet? I put our column-row in the first sheet. If you are interested, please feel free to edit the second sheet which shows 276 multiverse of BDA X AD cartesian product :)

https://docs.google.com/spreadsheets/d/1gbtGAA2YUaxXgHegiyoaa0WVKfPxnIgd7pxvXZABcIA/edit?usp=sharing

**Reply by hyunjimoon** [reply]
To be honest, I wish we could manage a blog together. Andrew's blog (platform where bayes ideas are generated and disseminated) [has many authors](https://statmodeling.stat.columbia.edu/authors/) (Bob, Dan, and Andrew's phd students) and many [sister blogs](https://statmodeling.stat.columbia.edu/blogs-i-read/). What do you think about changing the format of our goal from cookbook to cookblog? I have diffuse prior for its location, format etc and am waiting for your prior to either contract, constrain, compromise :)

I love every feature of your metasd blog (heritage, format), but if I may add one comment, I experienced graphical view (under the hood of [this blog](https://publish.obsidian.md/tolzul/%F0%9F%91%81%F0%9F%A7%A0%F0%9F%8C%B2%F0%9F%8F%AD+SimBasedCalib/def(SBC)%F0%9F%8C%80)) rendered by obsidian can benefit collaborative exploration. 
![image](https://user-images.githubusercontent.com/30194633/202908330-26819311-31f9-41a9-b87f-b5310edae41d.png)

**Comment by hyunjimoon** [comment]
from https://github.com/Data4DM/BayesSD/blob/master/DailyDigest/Moon22_ActionableWorkflow.pdf
![image](https://user-images.githubusercontent.com/30194633/203639744-4c993a5e-0753-4697-9294-5bb4d446a669.png)

**Comment by hyunjimoon** [comment]
First cut of auto-hierarchy stanify interface is as below in commit https://github.com/Data4DM/stanify/commit/81c4c09224347eab279764a4b993ea889c2a6c94:
```
setting = {
    "est_param_names": ("prey_birth_frac", "pred_birth_frac"),
    "hier_est_param_names": ("pred_birth_frac", ), #need ,) to be list downstream  # chosen among "est_param_names"
    "target_simulated_vector_names": ("prey", "predator"),
    "driving_vector_names": ("process_noise_uniform_driving"),
    "model_name": "S3R4",
}

precision ={
    "S": 3, # # of draws from prior
    "M": 10, # # of draws from posterior (# of chains * # of draws from each chain)
    "N": N, # # of observation
    "Q": len(setting['target_simulated_vector_names']), # # of target_simulated_stock
    "R": 4, # # of subgroups for hierarchical Bayes
    "time_step": .03,
    "integration_times": np.arange(0, N) * .03 + 0.01,  # initial_time < integration_times, .03 need not be time_step, but length should be N
}

numeric = {
    "process_noise_uniform_driving": np.random.uniform(low=-.5, high=.5, size = precision['N']),
    'process_noise_scale': 0.1
}

prior = {
    ("prey_birth_frac", "normal", 0.8, 0.08, 0),
    ("pred_birth_frac", "normal", 0.05, 0.005, 0),
    ("m_noise_scale", "normal", .01, .001, 0)
}

idata_kwargs = dict(
    prior_predictive=["prey_obs", "predator_obs"],
    posterior_predictive=["prey_obs_post", "predator_obs_post"],
    log_likelihood={
        "loglik": "loglik"
    },
    coords={
        "time": [n for n in range(precision['N'])],
        "stock": [q for q in range(precision['Q'])],
        "region": [r for r in range(precision['R'])]
    },
    dims={
        'initial_outcome': ["stock"],
        'integrated_result': ["time", "stock"],
        'prey': ["time"],
        'process_noise': ["time"],
        'predator': ["time"],
        "prey_obs": ["time"],
        "predator_obs": ["time"],
    }
)
```

Discussion needed on naming (left is Tom's right is Angie's word)
- downsampling, N
- ? , posterior predictive (controversy exists on predictive)     
```
prior_predictive=["prey_obs", "predator_obs"],
posterior_predictive=["prey_obs_post", "predator_obs_post"]
```

**Reply by hyunjimoon** [reply]
downsampling  going with decimating!

**Comment by hyunjimoon** [comment]
### in
- "below is .xml syntax. can you make a markdown table with parameter name as a row, equation (relation with other variables), unit as columns?"  "can you group this using the attached parameter classification?" 
-  🗄️-ed sd model's text [sd(💰project).md](https://github.com/user-attachments/files/17831255/sd.project.md)

- 🗄️ typifying parameter table in https://github.com/Data4DM/BayesSD/discussions/7#discussion-4388229, 

---

###  out

**Structural Parameters - Physical**
| Parameter | Equation/Value | Unit |
|-----------|---------------|------|
| Threshold RD&D | 1e+06 | $ |
| Founder's Basis | 1e+06 | $ |
| TIME STEP | 0.125 | Year |
| Potential Value | 2e+07 | $ |
| Investment Time | 2 | Year |

**Structural Parameters - Behavioral**
| Parameter | Equation/Value | Unit |
|-----------|---------------|------|
| Favorability of Terms to Founder | -0.3 | dmnl |
| Weight to Founder Basis | 2^Favorability of Terms to Founder | dmnl |
| Learning Coeff | 0.7 | dmnl |
| Min Burn Rate | 0.1 | fraction/Year |
| Burn Growth Rate | 0.5 | fraction/Year |
| Relative Investment Scale | 10 | dmnl |
| Threshold Knowledge | 4 | dmnl |

**State Variables and Flow Rates**
| Parameter | Equation/Value | Unit |
|-----------|---------------|------|
| Cash | INTEG(Investment - RD&D Expenditures, Founder's Basis) | $ |
| Cum RD&D | INTEG(RD&D Expenditures, 0) | $/Year |
| Knowledge | INTEG(Learning, 1) | dmnl |
| Investor Basis | INTEG(Investment, 0) | $ |

**Derived Variables**
| Parameter | Equation/Value | Unit |
|-----------|---------------|------|
| Investor Share | Investor Basis/(Founder's Basis×Weight to Founder Basis + Investor Basis) | fraction |
| Founder Value | Valuation×(1-Investor Share) | $ |
| Investor Value | Valuation×Investor Share | $ |
| Learning | Learning Coeff×Fractional Increase Expend×Knowledge | fraction/Year |
| Fractional Increase Expend | RD&D Expenditures/(Cum RD&D + Threshold RD&D) | fraction/Year |
| Investment | Founder's Basis×Relative Investment Scale×PULSE(Investment Time,TIME STEP)/TIME STEP | $/Year |
| RD&D Expenditures | Cum RD&D×Burn Growth Rate×(Cash/(Cash+Cum RD&D)) + Cash×Min Burn Rate | $/Year |
| Valuation | Cash + Potential Value×(Knowledge/Threshold Knowledge)/(1+EXP(-10×(Knowledge/Threshold Knowledge-1))) | $ |

Note: I've added two additional categories ("State Variables and Flow Rates" and "Derived Variables") since some parameters didn't strictly fit into the provided classification. The noise and prior parameters weren't clearly identifiable in the given model equations.


---
continue from https://poe.com/s/bL3yht0DQvk1J1voOpRI

**Reply by hyunjimoon** [reply]
Q1. meaning of ""Cum RD&D"*Burn Growth Rate*(Cash/(Cash+"Cum RD&D"))+Cash*Min Burn Rate"
Q2. how can cash with $ unit be accumulated to cum R&D with $/year unit? 
Q3. logic behind `burn growth rate` (.5) and `min burn rate` (.1) & both units are fraction/Year? 

<img width="707" alt="image" src="https://github.com/user-attachments/assets/880d1115-645f-43de-a619-d0fcc8b532a9">

**Reply by hyunjimoon** [reply]
would like to get feedback on, using [Synthesizing Model Parameters from Existing Tables cld](https://claude.ai/chat/0db50e34-c30b-444e-85e0-f545a625349b)

<img width="1257" alt="image" src="https://github.com/user-attachments/assets/97f1cf45-9ea8-4eda-b762-8e6487d7ada0">


| Type | Name | Formula/Value | Unit | Meaning | Process Role |
|------|------|---------------|-------|----------|--------------|
| 📊 Stocks |||||
|| 👁️ Observable | INTEG(diffusing_commitment - testing, 100) | unit | Raw information reservoir | Input source |
|| 🧠 Perceptual_Programmatic_Theory | INTEG(testing - implementing - decaying_perception, 50) | unit | Processed understanding | Knowledge transformation |
|| 🤜 Commitment | INTEG(implementing - diffusing_commitment - decaying_commitment, 0) | dmnl | Action decisions | Output state |
| 🔄 Flows |||||
|| 📍 testing | Theory/Time_to_test | unit/Week | 👁️→🧠 Processing raw data | Observation to perception |
|| 📈 implementing | Commitment/Time_to_implement | unit/Week | 🧠→🤜 Converting theory to action | Perception to action |
|| 💫 diffusing_commitment | Commitment/Time_to_diffuse | dmnl/Week | 🤜→👁️ Action impact on environment | Feedback loop closure |
| ⏱️ Time Parameters |||||
|| Time_to_test | 0.5 | Week | Sampling cost | Processing delay |
|| Time_to_implement | Time_to_test * implement_to_test_time_cost | Week | Action cost | Implementation delay |
|| Time_to_diffuse | 2 | Week | Market response time | Feedback delay |
| 🎚️ Decay Rates |||||
|| 📉 mu_p | 0.1 | dmnl/Week | Theory obsolescence rate | Knowledge deterioration |
|| 📉 mu_c | 0.1 | dmnl/Week | Action effectiveness decay | Impact diminishing |
| ⚙️ Constants |||||
|| implement_to_test_time_cost | 10 | dmnl | Action/perception cost ratio | Resource allocation |

Flow Logic:
1. Observable (👁️) → testing → Perceptual Theory (🧠) [Meaning construction]
2. Perceptual Theory (🧠) → implementing → Commitment (🤜) [Planning/action]
3. Commitment (🤜) → diffusing → Observable (👁️) [Environmental feedback]
4. Each stock has its own decay process (mu_p, mu_c) representing entropy

[test2implement.mdl.zip](https://github.com/user-attachments/files/17939408/test2implement.mdl.zip)
